File: ros2_ws/src/my_cpp_pkg/src/add_two_ints_client.cpp
#include <string_view>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using Srv_msg = example_interfaces::srv::AddTwoInts;
using Request = example_interfaces::srv::AddTwoInts_Request;
using Response = example_interfaces::srv::AddTwoInts_Response;

namespace {
std::string_view sv_name{"add_two_ints_client"};
}  // namespace

class AddTwoIntsClientNode : public rclcpp::Node {
 public:
  AddTwoIntsClientNode() : Node(sv_name.data()) {
    RCLCPP_INFO(this->get_logger(), "%s has started...", sv_name.data());

    thread1 = std::thread(
      std::bind(&AddTwoIntsClientNode::callAddTwoIntsService,
      this, 1, 2));
  }
  ~AddTwoIntsClientNode() {
    if(thread1.joinable()) {
      thread1.join();
    }
  }
  void callAddTwoIntsService(int a, int b) {
    client_ = this->create_client<Srv_msg>("add_two_ints");
    while(!client_->wait_for_service(std::chrono::seconds(1))) {
      RCLCPP_WARN(this->get_logger(), "Waiting for service...");
    }
    auto request = std::make_shared<Request>();
    request->a = 5;
    request->b = 5;

    auto future = client_->async_send_request(request);

    try {
      auto response = future.get(); // Blocking, spin(node) never runs
      RCLCPP_INFO(this->get_logger(), "%d + %d = %ld", a, b, response->sum);
    } catch(const std::exception &e) {
      RCLCPP_ERROR(this->get_logger(), "Service call failed.");
    }



  }

 private:
 std::thread thread1;
 std::vector<std::thread> threads_;
 rclcpp::Client<Srv_msg>::SharedPtr client_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<AddTwoIntsClientNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
from functools import partial
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

node_name = 'add_two_ints_client'

class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__(f'{node_name}')
        self.get_logger().info(f'{node_name} has started...')
        self.client_ = self.create_client(AddTwoInts, 'add_two_ints')
        self.call_add_two_ints_server(6,7)
        self.call_add_two_ints_server(6,8)
        self.call_add_two_ints_server(6,9)
    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, 'add_two_ints')
        # Check if server is up
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server...")
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future = client.call_async(request)
        future.add_done_callback(partial(self.callback_add_two_ints, a=a, b=b))
    def callback_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(f'{a} + {b} = {response.sum}')
        except Exception as e:
            self.get_logger().error('Service called failed: {e}')

def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()